### ISSUE TITLE
WB Modified Date not saved after upload

### ISSUE THREAD

**Issue by grantfitzsimmons** [specify/specify7#1806]
![image](https://user-images.githubusercontent.com/37256050/177856787-ceef1ba1-e13f-417e-9dd5-6b0d3ff004d0.png)

---

**Expected Behavior:**
* When a column within a specified table is linked to either `timestampModified` or `timestampCreated`, it should not be automatically replaced upon uploading new data. 
* In cases where a column is not associated with `timestampModified` or `timestampCreated`, the system should autonomously allocate both fields with the current timestamp. 
* When new records are generated in other tables during the same upload process, if no column is directly linked to `timestampModified` or `timestampCreated`, the system should automatically assign both fields with the current timestamp.
* `timestampModified` or `timestampCreated` should be assigned the current timestamp in all cases during WorkBench uploads **except** when a column is mapped to that field where that row contains a value

**Comment by maxxxxxdlp** [comment]
Back-end overwrites the `timestampModified` (and I would imagine `timestampCreated` too) field when a new resource is created.

This makes data migrations harder.

**Comment by grantfitzsimmons** [comment]
From CSIRO:

> We wanted to escalate the fix of github ticket #1806 ( as we didn't realise just how damaging this issue is to our collections. The fact that the system is not updating the datemodified field means we are losing data integrity as we cannot run any audits on when specimen records are being modified by users.
>
> From our testing of this any time we upload something to the workbench that is linked to a collection object (determination, preparation, voucher relationship etc.) the creation event is logged in the audit log tables under that specific table, but there is no row in the audit log that shows that the related collection object was modified. Both the datemodified and version fields for the collection object are not updated. If you use a form to manually create/edit a determination then the audit log correctly records the event for the determination table and the collection object table, and the later has the correct datemodified and version values updated.
>
> This bug is a breach of FAIR data policy since these edits are no longer "Findable" by our users. All of our collections need to report on their activity, and that includes reporting newly databased specimens within a time period, as well as edits/additions to any specimen.

**Comment by grantfitzsimmons** [comment]
This is considered SUPER important to them, so we need to discuss this to assess the difficulty in our development meetings.

**Comment by melton-jason** [comment]
This behavior is likely due to the backend setting a field-specific option on Django (`auto_now`) which will ALWAYS use the current timestamp whenever the record is saved.  

https://github.com/specify/specify7/blob/c1f47c3c331c804853b9cba0d4c6f13d51185e2a/specifyweb/specify/build_models.py#L41-L42

> Automatically set the field to now every time the object is saved. Useful for “last-modified” timestamps. Note that the current date is always used; it’s not just a default value that you can override.

https://docs.djangoproject.com/en/3.2/ref/models/fields/#django.db.models.DateField.auto_now

**Comment by acwhite211** [comment]
In the database, these fields are defined like this:
```sql
`TimestampCreated` datetime NOT NULL,
`TimestampModified` datetime DEFAULT NULL,
```

In the Django model for these fields we have:
```python
timestampcreated = models.DateTimeField(auto_now_add=True)
timestampmodified = models.DateTimeField(auto_now=True)
```

Django is always updating timestampmodified when the record is changed, and timestampcreated when the record is created.

I thinking it would be a good idea to just temporarily suspend this behavior, like so:
```python
field.auto_now = False
field.auto_now_add = False
 
# ... make updates

record.save()

field.auto_now = True
field.auto_now_add = True
```

I suppose we need to define exactly when and when not to have these fields get automatically updated.  I feel that having this functionality is useful most of the time, so it might be better to just suppress this behavior in certain cases.

___

If I'm understanding this correctly, CSIRO wants to modify a row in a table, have the timestampmodified get updated, and then update timestampmodified of the related rows in other tables?  Is that every table that has a relationship to the original table?  Is it just a one relationship hop, or should we also be updating timestampmodified relationships of relationships?  

From a relational database perspective, this seems like a bad idea.  In their example, the collection object wasn't truly modified, only the determination it has a relationship to was.  Someone doing an audit could do a query to see that this collection object's determination was modified at this specific time.  So in this example, they are just looking in the wrong place for the modified date. 

They mentioned that if they used the form view to edit the determination, instead of the workbench, then the collection object timestampmodified gets updated.  Is that logic on the front-end that determines which related records to update?

If we decide to manually modify timestampmodified, we need to define exactly which relationships of a table get their timestampmodified field overwritten, and which ones don't.

It seems to me that this is just a misunderstanding of an RDS vs spreadsheet mindset.

**Comment by acwhite211** [comment]
I talked with Jason and Grant.  The solution we settled on is to temporarily suspend `auto_now` and `auto_now_add` Django behavior when the API request to update a record includes the `timestampmodified` or `timestampcreated` fields.

**Comment by acwhite211** [comment]
Now that I think of it, disabling the auto_now temporarily might not be the best idea because the workbench upload runs on the celery worker, so it would change the Django model while running in the background.  So if the upload took a while, the Django model's auto_now will still be modified while other requests come in, not ideal.

**Comment by acwhite211** [comment]
We need to allow for the user to map columns to the `timestampmodified` and `timestampcreated` fields when creating an upload plan.  There must have been a front-end change sometime after this issue was reported where we disabled the option to map to these fields, even when checking 'display hidden fields'.  That change was probably a good idea, but will need to be reversed for this issue.

**Comment by acwhite211** [comment]
If we remove the `auto_now` from `timestampmodified`, and `auto_now_add` from `timestampcreated`, then we could maybe override the save function of the Django models.
```python
class SpTimestampedModel(models.Model):
    """
    SpTimestampedModel(id, timestampcreated, timestampmodified)
    """

    # Fields
    timestampcreated = models.DateTimeField(db_column='TimestampCreated')
    timestampmodified = models.DateTimeField(db_column='TimestampModified')

    class Meta:
        abstract = True

    def save(self, *args, **kwargs):
        timestamp_override = kwargs.pop('timestamp_override', False)

        if not timestamp_override:
            if not self.id:
                self.timestampcreated = timezone.now()
            
            self.timestampmodified = timezone.now()
        
        super().save(*args, **kwargs)
```

**Comment by CorinnaPaeper** [comment]
Hi @acwhite211  and @grantfitzsimmons,

Just to clarify, we are not asking for a change to the behaviour of when the TimestampModified field gets updated in the CO table. It has always been the case that the version number and TimestampModified of the related CO record was updated when any linked table to the CO table was created/modified via the forms. This behaviour works as intended (and as expected by us), as a change in something like the determiner ultimately is a change of the CO information.

Your suggestion of checking the related table for a TimestampModified is not really user friendly as you either have to know exactly which related table of the record was changed, or you have to add all TimestampModified fields of EVERY related table to the query. As the query rows act like an AND statement, you will not be able to put any useful conditions on a query like that (which means any analysis like this will have to be completed outside of the Specify UI).

It makes complete sense for us that the version and TimestampModified of a CO is updated when a related table is edited. It means we can run an easy query that identifies what has changed and allows our collections to do their annual audits of database activity. Right now doing the same action in the Workbench as in the forms DOES NOT yield the same result. 

We hope this clarifies our issue. 
The CSIRO DMT

**Comment by grantfitzsimmons** [comment]
Hi @CorinnaPaeper,

This issue (#1806) is specifically about the WorkBench setting the  `timestampModified` value to be the time of upload regardless of what is mapped in the upload plan. The solution to this issue is to have Specify preserve the value from that column once the upload is complete.

> <img width="522" alt="image" src="https://github.com/specify/specify7/assets/37256050/6eaae0e4-81b3-48bb-b4dd-bcf3305ab42f">
> 
> Despite a column being mapped to `timestampModified` and containing a value of `8/23/2019`, Specify automatically sets the timestamp modified to the current date (it was `07/07/2022` when this was initially reported).

Based on the discussions I've had with the developers about resolving this issue in particular, there should be no change to the default behavior of assigning `timestampCreated` or `timestampModified` values **unless** the user explicitly maps a field in a WorkBench upload to one of those fields in any table.

If there is nothing mapped to one of those fields, it should use the existing behavior to assign those values. 

---

Based on your [comment](https://github.com/specify/specify7/issues/1806#issuecomment-1992918309) and the Asana ticket review, it appears you may be asking for help with a different issue than what #1806 aims to address.

Rather than the problem being that you cannot manually set the `timestampModified` from a value in an import spreadsheet, it sounds like you need the WorkBench to automatically set the `timestampModified` of the Collection Object associated when a dependent record is uploaded and associated with it (determination, preparation, attribute, etc.). 

Is that so? If yes, I will need to open a new issue to track this separately.

Thank you and the team for taking the time to reply!

**Comment by maxpatiiuk** [comment]
Regarding making timestamp field mappable in the workbench, they are excluded in this file at the moment:
https://github.com/specify/specify7/blob/production/specifyweb/frontend/js_src/lib/components/DataModel/schemaOverrides.ts

Modifying this line to say "optional" instead of "hidden" might be all that's needed on the front-end:

https://github.com/specify/specify7/blob/aa434e5d24f9b88b3a2c2d892d8619f5d35e815e/specifyweb/frontend/js_src/lib/components/DataModel/schemaOverrides.ts#L187
